src/util/newconfig/ConfigDefinition.cpp

```diff
@@ -101,113 +108,102 @@ static ClioConfigDefinition ClioConfig = ClioConfigDefinition{
-            return ObjectView{prefix, *this};
-
-        if (mapKey.starts_with(prefixWithDot) && std::holds_alternative<Array>(mapVal)) {
-            ASSERT(std::get<Array>(mapVal).size() > idx, "index provided is out of scope");
+        auto const hasPrefix = mapKey.starts_with(prefixWithDot);
+        if (idx.has_value() && hasPrefix && std::holds_alternative<Array>(mapVal)) {
+            ASSERT(std::get<Array>(mapVal).size() > idx.value(), "Index provided is out of scope");
+            // we want to support getObject("array") and getObject("array.[]"), so we check if "[]" exists
+            if (!prefix.contains("[]"))
+                return ObjectView{prefixWithDot + "[]", idx.value(), *this};
             return ObjectView{prefix, idx.value(), *this};
         }
+        if (hasPrefix && !idx.has_value() && !mapKey.contains(prefixWithDot + "[]"))
+            return ObjectView{prefix, *this};
     }
-    throw std::invalid_argument("Key is not found in config");
+    ASSERT(false, "Key {} is not found in config", prefixWithDot);
+    return ObjectView{"", *this};
 }
 
 ArrayView
 ClioConfigDefinition::getArray(std::string_view prefix) const
 {
-    std::string key = std::string(prefix);
-    if (!prefix.contains(".[]"))
-        key += ".[]";
+    auto key = addBracketsForArrayKey(prefix);
 
     for (auto const& [mapKey, mapVal] : map_) {
-        if (mapKey.starts_with(key))
+        if (mapKey.starts_with(key)) {
+            ASSERT(
+                std::holds_alternative<Array>(mapVal), "Trying to retrieve Object or ConfigValue, instead of an Array "
+            );
             return ArrayView{key, *this};
-        ASSERT(!mapKey.starts_with(prefix), "Trying to retrieve an Array when value is an object");
+        }
     }
-    throw std::invalid_argument("Key is not found in config");
+    ASSERT(false, "Key {} is not found in config", key);
+    return ArrayView{"", *this};
+}
+
+bool
+ClioConfigDefinition::contains(std::string_view key) const
+{
+    return map_.contains(key);
+}
+
+bool
+ClioConfigDefinition::hasItemsWithPrefix(std::string_view key) const
+{
+    auto it = std::ranges::find_if(map_, [&key](auto const& pair) { return pair.first.starts_with(key); });
+    return it != map_.end();
 }
 
 ValueView
 ClioConfigDefinition::getValue(std::string_view fullKey) const
 {
-    if (map_.contains(fullKey) && std::holds_alternative<ConfigValue>(map_.at(fullKey))) {
+    ASSERT(map_.contains(fullKey), "key {} does not exist in config", fullKey);
+    if (std::holds_alternative<ConfigValue>(map_.at(fullKey))) {
         return ValueView{std::get<ConfigValue>(map_.at(fullKey))};
     }
-    ASSERT(
-        !map_.contains(fullKey) && std::holds_alternative<Array>(map_.at(fullKey)), "value of key is not Config Value."
-    );
+    ASSERT(false, "Value of key {} is an Array, not an object", fullKey);
+    return ValueView{ConfigValue{}};
```

Comment:
This can never be reached, right? there is `std::unreachable` for that. (https://compiler-explorer.com/z/1EGsMTaz5)

Reply:
Keep in mind that if for some reason the ASSERT does not work and `std::unreachable` is actually reached it will lead to UB and a SEGFAULT in practice.